### PR TITLE
Improve phase teleportation

### PR DESCRIPTION
**Pull request by ChenZhao44**
Try to find out the reason for the higher T-count in phase teleportation.
Still working. Please do not merge!

### CODE REVIEW

File: src/simplify.jl [hunk 1]
```diff
@@ -15,10 +15,16 @@ end
...
+    max_iter = 20
+    i = 1
     matches = match(r, zxd)
-    while length(matches) > 0
+    while length(matches) > 0 && (i <= max_iter || r ∉ (Rule{:p2}(), Rule{:p3}()))
         rewrite!(r, zxd, matches)
         matches = match(r, zxd)
+        i += 1
+    end
+    if i > max_iter && r in (Rule{:p2}(), Rule{:p3}())
```
Comment (Roger-luo) [inline src/simplify.jl:26]: should this warning be inside the while loop?

File: src/rules.jl [hunk 2]
```diff
@@ -618,13 +642,14 @@ function check_rule(::Rule{:p2}, zxg::ZXGraph{T, P}, vs::Vector{T}) where {T, P}
...
-            (qubit_loc(zxg, v1) !== nothing || nqubits(zxg) == 0)
+            (degree(zxg, v1)) > 1 && (rem(phase(zxg, v1), 1//2) != 0) &&
+            length(neighbors(zxg, v1)) > 1
             if v2 in neighbors(zxg, v1)
                 if spider_type(zxg, v2) == SpiderType.Z && is_interior(zxg, v2) &&
-                    (phase(zxg, v2) == 0 || phase(zxg, v2) == 1) &&
-                    (qubit_loc(zxg, v2) !== nothing || nqubits(zxg) == 0)
-                    return true
+                    (phase(zxg, v2) in (0, 1))
+                    if all([length(neighbors(zxg, u)) > 1 for u in neighbors(zxg, v2)])
```
Comment (Roger-luo) [inline src/rules.jl:650]: no need to use `[]` just 
```suggestion
                    if all(length(neighbors(zxg, u)) > 1 for u in neighbors(zxg, v2))
```

Comment (Roger-luo) [reply src/rules.jl:650]: there is a few other places like this.

**Review by Roger-luo — COMMENTED** [review]
looks good in general. A few places need to polish tho.

File: src/simplify.jl (hunk 1 above)
Comment (Roger-luo) [reply src/simplify.jl:26]:
```julia
const MAX_ITERATION = Ref{Int}(1000)
```

**Comment by ChenZhao44** [comment]
Applying Rule p2 and p3 on the boundary spiders may generate untrackable spiders. And it will cause phase teleportation to fail. So I stopped working on it. 
This branch is ready to be merged now. @Roger-luo

**Pull request merged**